src/realm/util/future.hpp

```diff
@@ -666,12 +666,12 @@ class REALM_NODISCARD future_details::Future {
     /**
      * This ends the Future continuation chain by calling a callback on completion. Use this to
      * escape back into a callback-based API.
+     *
+     * The callback must not throw since it is called from a noexcept context.
      */
     template <typename Func> // StatusOrStatusWith<T> -> void
     void get_async(Func&& func) && noexcept
     {
```

Comment:
So I think you want something like `static_assert(std::is_void_v<std::invoke_result_t<Func, StatusOrStatusWith<FakeVoidToVoid<T>>>>);` to ensure that `Func` isn't returning a non-void result. The old code wasn't actually doing that correctly because (quoting [cppreference](https://en.cppreference.com/w/cpp/types/is_invocable):
> If `R` = `void`, the result can be any type.

The mongo repo version of this code handled this is using its own helper to handle this and a few other annoyances around `is_invocable`. Sorry that this crap is so complicated.

Reply:
Fixed.